packages/ui/components/ui/json-viewer.tsx

```diff
@@ -1,30 +1,129 @@
 import type { JsonObject, JsonValue } from '@bufbuild/protobuf';
 import { JsonView } from 'react-json-view-lite';
+import { cn } from '../../lib/utils';
+import { CopyToClipboardIconButton } from './copy-to-clipboard-icon-button';
+import { DoubleArrowDownIcon, DoubleArrowUpIcon } from '@radix-ui/react-icons';
+import { Button } from './button';
+import { useCallback, useState } from 'react';
+
+const collapseLevel = 3;
+
+const objectDepth = (o: JsonValue): number =>
+  o && typeof o === 'object' ? 1 + Math.max(-1, ...Object.values(o).map(objectDepth)) : 0;
+
+const objectLength = (o: JsonValue): number =>
+  o && typeof o === 'object' ? Object.entries(o).length : 0;
 
 export const JsonViewer = ({ jsonObj }: { jsonObj: JsonObject | JsonValue[] }) => {
+  const [expandAll, setExpandAll] = useState(false);
+
+  const shouldExpandNode = useCallback(
+    (level: number, value: JsonValue, field?: string) => {
+      if (expandAll) return true;
+      if (
+        // expand empty, so they can be collapsed by css
```

Comment:
Not sure I understand this — collapsed by CSS?

Reply:
instead of ever rendering an empty object as `field: { ••• }` and causing the user to click it and see nothing, all empty objects are expanded. then css can render it compactly as `field: { }`

when i wrote the comment i think i didn't have 'collapsed' showing as inline yet, so a collapsed empty would take up multiple lines.